### ISSUE TITLE
consistent short flags/keys for tree and flat mode

### ISSUE THREAD

**Issue by simonmichael** [simonmichael/hledger#1286]
I'm thinking we should 

1. assign -T and -F as shortcuts for --tree and --flat in accounts/balance reports. It's a common enough flag in practice that I think the short flags will be handy.
2. assign the T and F keys for choosing tree mode and flat mode in hledger-ui. This would mean changing the F key's current function (toggle future mode). That's unfortunate but I think acceptable, because F actually used to toggle between tree and flat mode until 1.12, and even now I still find myself trying to use it for that. Also I think (though haven't tested) that having a key per mode, rather than a one-key toggle, will be a usability improvement for this important function.

An open question is what key to use for toggling hledger-ui's future mode ? There aren't many good choices available. Some options:

- O, as in fOrecast. Dosn't correspond with the meaning of hledger's -O flag, but arguably that feature will never be needed in hledger-ui. It is a valuable sort of letter though.
- Z, because it's available, distinct, and uh could mean "jump to the end of things". Also a valuable letter.
- start using lower case letters, such as f. That would be a first, and would prevent using those letters for something more cool (quick data entry, selecting bookmarks/predefined reports, or something). Also f doesn't correspond with hledger's -f.
- One of the punctuation symbols ? We may start to get into a cryptic unix utility/nethack sort of vibe. The shifted number keys are also potentially valuable for other things (defining bookmarks/custom reports..)
- Other ideas ?

**Comment by simonmichael** [comment]
FWIW here's what -F and -T mean in Ledger:
```
     --format FMT (-F)
              Use the given format string FMT to print output.

     --total EXPR (-T)
              Define a value expression used to calculate the total in reports.
```

**Comment by simonmichael** [comment]
A setback. -T enables balance reports' Total column, and I use it often.
```
  -T --row-total             show a row total column (in multicolumn reports)
```

**Comment by simonmichael** [comment]
New plan: I think `-t`/`t` and `-l`/`l` are a decent choice. We do already use lower case letters in hledger-ui, so it's fine to use more, especially for common operations. And tree and list are a natural pairing. It would be convenient if --flat had been --list from the start. It's not worth changing that flag and all our docs, and probably not worth adding a --list alias, but `l`/`flat`/`list` are related enough it should still be fairly comfortable/mnemonic.

**Comment by simonmichael** [comment]
`t` is already used in hledger-ui for "set report period to today". No problem, at this point I'm ready to break stuff. I'll change that one to `T`.
*...hack hack document document document...*
`l` is also used, for "move right".
*Arggggggh...*

**Comment by simonmichael** [comment]
Use `t` to toggle I guess. At least it matches one of the cli flags. It's that or drop the vi keys.

**Comment by simonmichael** [comment]
I dropped the vi movement keys. Arguments:

- I don't know of anyone using them in hledger-ui
- four unmodified letter keys are too valuable to give up, considering there are two other sets of movement keys (arrows, emacs control keys)
- nobody is likely to run hledger-ui on a keyboard that has neither arrow nor control keys

If you disagree, please comment here or in chat!

**Comment by simonmichael** [comment]
Or, which of the following would you prefer ?

1. drop vi keys and use t and l keys to choose tree or list mode
2. keep vi keys and use t to toggle tree/list mode
3. <something else>

**Comment by cfellinger** [comment]
Hi all,

as user interface design is the thing I enjoyed most in my working career:

On Fri, Jul 17, 2020 at 04:09:59PM -0700, Simon Michael wrote:
> Or, which of the following would you prefer ?
> 
> 1. drop vi keys and use t and l keys to choose tree or list mode
> 2. keep vi keys and use t to toggle tree/list mode
> 3. (something else)

Those of us that prefer vi are just like those of us that prefer emacs,
they *vehemently* like that one to the point their fingers can't help but
use those keys for movement.  Moreover, the vi party has well founded
sentiments about not having to move their fingers from the home row.

Muscle memory is very persistent, e.g. I always type q to quit the
help menu!  I know it makes me restart and seek again, but I can't
help myself and stubernly type it again and again.  So be gently to
the (few) remaining vi die-hards and hence use t to toggle tree/list.


More on the q to quit help:
--------------------------

Most keys don't work in help mode, but q, cursor back and p/m/i do.
The funny thing is that the extende help provided by p/m/i do need a q
to bring you back to hledger-ui.  So:

ceterum cenceo "q should get you out of help mode back into hledger"

-- 
groetjes, carel

**Comment by simonmichael** [comment]
On Fri, Jul 17, 2020, at 5:23 PM, Carel Fellinger wrote:
> On Fri, Jul 17, 2020 at 04:09:59PM -0700, Simon Michael wrote:
> 
> > Or, which of the following would you prefer ?
> > 
> > 1. drop vi keys and use t and l keys to choose tree or list mode
> > 2. keep vi keys and use t to toggle tree/list mode
> > 3. (something else)
> 
> Those of us that prefer vi are just like those of us that prefer emacs,
> they *vehemently* like that one to the point their fingers can't help but
> use those keys for movement. Moreover, the vi party has well founded
> sentiments about not having to move their fingers from the home row.
> 
> Muscle memory is very persistent, e.g. I always type q to quit the
> help menu! I know it makes me restart and seek again, but I can't
> help myself and stubernly type it again and again. So be gently to
> the (few) remaining vi die-hards and hence use t to toggle tree/list.

Aha, there's one! Thanks for the input Carel.

I knew it was Wrong, but I had done too much work updating help texts etc. to stop. I expect it'll be reverted soon.

> More on the q to quit help:
> --------------------------
> 
> Most keys don't work in help mode, but q, cursor back and p/m/i do.
> The funny thing is that the extende help provided by p/m/i do need a q
> to bring you back to hledger-ui. So:
> 
> ceterum cenceo "q should get you out of help mode back into hledger" 

Hmm, interesting. LEFT or ESC are supposed to be the way out of help, as it says at the top. And I feel like such an important key as q should do one thing, reliably. But it's true that you also have to use it to get out of the help pager, man, or info. Perhaps q should be another way out, and not quite so powerful ?

Related, should exiting the pager/man/info return to the help dialog ? Or is it best how it is (closes the help dialog too) ?

FYI I also tweaked the ESC key so that it clears any report period filter, when I couldn't remember how to do that. Sometimes I want to clear some UI state but not all of it, but for now ESC does a full reset, ie "confused, start over!"

**Comment by cfellinger** [comment]
Hai again,

On Fri, Jul 17, 2020 at 06:00:55PM -0700, Simon Michael wrote:
> On Fri, Jul 17, 2020, at 5:23 PM, Carel Fellinger wrote:
> > More on the q to quit help:
> > --------------------------
> > 
> > Most keys don't work in help mode, but q, cursor back and p/m/i do.
> > The funny thing is that the extende help provided by p/m/i do need a q
> > to bring you back to hledger-ui. So:
> > 
> > ceterum cenceo "q should get you out of help mode back into hledger" 
 
> Hmm, interesting. LEFT or ESC are supposed to be the way out of

hmm, LEFT works, but ESC is very disruptive as a way out of help :(

> help, as it says at the top. And I feel like such an important key
> as q should do one thing, reliably. But it's true that you also have
> to use it to get out of the help pager, man, or info. Perhaps q
> should be another way out, and not quite so powerful ?

You see, the tricky thing about UI is that rigor kills it,
but consistent logic/behaviour *is* essential to make it swing.

So, yes being able to quit the program is important, but appreciating
the difference between the help menu and pager/man/info mode in that
respect, that one is part of the program and the other is an external
invocation is tricky.  How can I know that the help menu isn't
provided by a separate program or that the pager is?

But I do see them all as part of the help system, a special mode
separated from the normal program flow to get meta info.  So I
think it's perfectly sane to expect q to quit that meta level.
It does so, just not for the builtin help menu.

Here is your change to correct that and
establish a consistent UI :)

 
> Related, should exiting the pager/man/info return to the help dialog?
> Or is it best how it is (closes the help dialog too) ?

It did surprised me slightly, but only now checking the behaviour in
the help menu before commenting on it.  The help menu provides limited
help and --in contrast-- the pager/man/info help systems give the same
and much more. So it seems not likely that you want to return to the
simple help menu after having explored the full docs.  Hence the
current behaviour is fine.


> FYI I also tweaked the ESC key so that it clears any report period
> filter, when I couldn't remember how to do that. Sometimes I want to
> clear some UI state but not all of it, but for now ESC does a full
> reset, ie "confused, start over!"

Again something that surprised me the first time around, but it *is*
reassuring to know a way to start afresh without having to restart !
The question is what is the most confortable level to "start afresh".

I did a small check, and flat/tree seems to be included.  To me,
flat/tree is a very global preference.  Not related to filtering at
all.  And on top of that, when I specify --tree at the command line I
expect ESC to honor that.

To me that gives a hint for the proper behaviour of ESC:

  reset the ui tweaks done within the program reverting
  to the command line flags and the program defaults.

The thing that bugs me is whether to go back to the initial screen, or
stay at the screen and the position you're on.  I'm rather peculiar on
not disrupting the work flow.  So when would I press ESC?  Not likely
because I'm lost in the screen hierarchy.  So having to go back to the
screen and the position I explicitly selected just to reset filters
seems very unpleasant to say the least.  Resetting flat/tree, date,
historical and UPC selections on the other hand fit perfectly well
with filter reset, especially when honoring the command line flags.

-- 
groetjes, carel

**Comment by hpdeifel** [comment]
As a data point: I use the vi keys all the time. Of course I can also see myself using the arrow keys, but hjkl is very much ingrained in my brain at this point and I always like UIs that support them.

**Comment by simonmichael** [comment]
Thanks all for your input. Normality has been restored! `t` will toggle tree mode.

And, `q` is now another way to exit the help dialog (so you may have to hammer on it twice to exit the app).

> To me that gives a hint for the proper behaviour of ESC:
> 
> > reset the ui tweaks done within the program reverting
> > to the command line flags and the program defaults.

Good point: agreed that reset should preserve any startup options/arguments.

> The thing that bugs me is whether to go back to the initial screen, or
> stay at the screen and the position you're on.  I'm rather peculiar on
> not disrupting the work flow.  So when would I press ESC?  Not likely
> because I'm lost in the screen hierarchy.  So having to go back to the
> screen and the position I explicitly selected just to reset filters
> seems very unpleasant to say the least.  Resetting flat/tree, date,
> historical and UPC selections on the other hand fit perfectly well
> with filter reset, especially when honoring the command line flags.

So I guess we're not 100% clear on what ESC should reset ? Sometimes I want more, sometimes less. Should it reset progressively more as you hammer on it one, two, tree times ? I thought that would be too confusing but maybe not.

**Comment by simonmichael** [comment]
Tweaked these keys:

- ESC resets the whole app to its initial state (as it appeared when you first ran it)
- DEL just resets the filters & toggles to the initial state (and no longer changes tree/list mode)

**Comment by cfellinger** [comment]
Hai,

great how this works,
a question, a few reactions, and a change for the better :)

On Sat, Jul 18, 2020 at 10:57:07AM -0700, Simon Michael wrote:
> > The thing that bugs me is whether to go back to the initial screen, or
> > stay at the screen and the position you're on.  I'm rather peculiar on
> > not disrupting the work flow.  So when would I press ESC?  Not likely
> > because I'm lost in the screen hierarchy.  So having to go back to the
> > screen and the position I explicitly selected just to reset filters
> > seems very unpleasant to say the least.  Resetting flat/tree, date,
> > historical and UPC selections on the other hand fit perfectly well
> > with filter reset, especially when honoring the command line flags.

> So I guess we're not 100% clear on what ESC should reset ? Sometimes
> I want more, sometimes less. Should it reset progressively more as
> you hammer on it one, two, tree times ? I thought that would be too
> confusing but maybe not.

No no, I don't like multi level escapes either.
Just one level, and reset all you like, but please stick to
the selected screen and position in that screen as mutch as possible.

To me that is the one thing I hate changing without me stipulating it.
Just like when toggling with UPC, it works much better for us now that
the on screen location is preserved (more or less) whilst toggling.

We use UPC to see how far we are with reconciling: pressing END or
BEGIN in between toggles gives us now the needed overview, thanks!

Similary, setting filters and such to me is a tool to find interesting
transactions.  The fact that currently clearing the filters will bring
me back to the primairy screen ruins all the fun of filtering: I am
again at lost as where to find that interesting transaction.

On the other hand, not resetting screen and position hardly places a
burdon upon the user. Navigating through the account hierarchy seems
such a basic skill, it likely is second nature to any hledger user.

So to sum up: one level of escape resetting UI to --flags and defaults.
But do leave screen & position as explicitly navigated to by the user.

-- 
groetjes, carel

**Comment by cfellinger** [comment]
Hai Simon,

sorry, I just mist this post.

On Sat, Jul 18, 2020 at 12:35:32PM -0700, Simon Michael wrote:
> Tweaked these keys:
 
> - ESC resets the whole app to its initial state (as it appeared when
> - you first ran it) DEL just resets the filters & toggles to the
> - initial state (and no longer changes tree/list mode)

That works for me.
Thanks a lot for making this program and increasing the fun to use it !

-- 
groetjes, carel

**Comment by hyiltiz** [comment]
> keep vi keys and use t to toggle tree/list mode
I'd very much like this. In fact, I've never noticed `hledger-ui` existed until I played with the source recently. I tried it out briefly, and apparently didn't feel the need to learn or look up anything as the interface was quite "intuitive" (for me, subjectively). On the other hand, if it wasn't vi-keys, I'd definitely have to look up things.
